### PR TITLE
refactored to read all options prior to setting up discretizations

### PR DESCRIPTION
**Pull request by ghammond86**

### CODE REVIEW

**Comment by bishtgautam** [comment]
@ghammond86 Can you describe the change? Are demos now not allowed to call `TDySetFromOptions()`?

**Comment by ghammond86** [comment]
Prior to this refactor, the initialization of TDycore many objects (discretization, matrices, vectors, etc.) was intermingled with the reading of options.  This poses a problem when command line options override default settings programmed into the demos.  The refactor ensures that all command line options are read and parsed prior to creating objects.  Demos may call TDySetFromOptions() prior to TDySetup(), but TDySetFromOptions() is automatically called within TDySetup().

**Comment by bishtgautam** [comment]
Why not force the demos to call only `TDySetup()`?

**Comment by ghammond86** [comment]
The demos do only call TDySetip().

Glenn

On Oct 9, 2020, at 10:17 PM, Gautam Bisht <notifications@github.com> wrote:

﻿

Why not force the demos to call only TDySetup()?

**Review by jedbrown — COMMENTED** [review]
Is this fixing the issue where fields were being created based on the default discretization?

**Comment by bishtgautam** [comment]
@ghammond86 In that case, let's get rid of `TDSetFromOptions()` from [tdycore.h](https://github.com/TDycores-Project/TDycore/blob/glenn/refactor-options/include/tdycore.h#L77) and the equivalent Fortran interfaces.

**Review by bishtgautam — CHANGES_REQUESTED** [review]
Remove TDSetFromOptions() from tdycore.h and the equivalent Fortran interfaces.

File: demo/steady/steady.c [hunk 1]
```diff
@@ -744,6 +743,13 @@ int main(int argc, char **argv) {
   ierr = DMCreateMatrix      (dm,&K ); CHKERRQ(ierr);
   ierr = TDyComputeSystem(tdy,K,F); CHKERRQ(ierr);
 
+  PetscViewer viewer;
+#if 0
+  ierr = PetscViewerASCIIOpen(PETSC_COMM_WORLD,"K.mat",&viewer); CHKERRQ(ierr);
+  ierr = MatView(K,viewer);
+  ierr = PetscViewerDestroy(&viewer); CHKERRQ(ierr);
+#endif
```
[the remark this answers is not in the document]
Comment (ghammond86) [reply demo/steady/steady.c:751]: PetscViewer is required by the subsequent viewers below the #if 0 block.  I removed the block to clean up the code.

File: src/tdycore.c [hunk 2]
```diff
@@ -551,53 +554,45 @@ PetscErrorCode TDySetFromOptions(TDy tdy) {
                           &flg); CHKERRQ(ierr);
   if (flg && (bctype != tdy->mpfao_bc_type)) { ierr = TDySetMPFAOBoundaryConditionType(tdy,bctype); CHKERRQ(ierr); }
 
+  ierr = PetscOptionsEnd(); CHKERRQ(ierr);
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode TDySetup(TDy tdy) {
+  // must come after TDySetFromOptions()
```
[the remark this answers is not in the document]
Comment (ghammond86) [reply src/tdycore.c:562]: This was a comment from an intermediate commit.  Squashing with rebase does not make this evident as the history is destroyed.  I have remove it.

File: demo/steady/steady.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ghammond86) [reply demo/steady/steady.c]: I believe that the PR resolves the issue of fields being created off the default discretization, but I would need an example problem to compare before and after.

**Review by bishtgautam — APPROVED** [review]

**Comment by bishtgautam** [comment]
@ghammond86 Why do the gold files need to be updated for TPF?

**Comment by ghammond86** [comment]
> @ghammond86 Why do the gold files need to be updated for TPF?

The TPF was using the MPFA_O matrix, which had extra columns (with zeros) than the standard star stencil.  During ILU factorization, those factored entries seemed to be tweaking the answer ever so slightly.  I ran comparisons with Jacobi (diagonal only) and direct LU preconditioning and the extra column and standard star stencil matrices matched.  What I do not understand is that the results for ILU (with extra columns and standard star) did not match Jacobi preconditioning or direct LU.  But I had to move on.  I trust the results as the matrices had identical values.

**Pull request merged**